### PR TITLE
Fix local file resolver logic with relative paths and site-packages

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
Change SitePackageResolver logic, such that it looks for `_package_/__init__.py` rather than rely on `dist-info` metadata.
Wire resolver back in global resolver chain
Update calling code

### Linked issues
#1202 

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests
- [ ] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: src/databricks/labs/ucx/source_code/files.py [hunk 1]
```diff
@@ -150,3 +150,29 @@ def resolve_import(self, path_lookup: PathLookup, name: str) -> MaybeDependency:
...
+class SitePackageResolver(BaseDependencyResolver):
+
+    def __init__(
+        self,
+        file_loader: FileLoader,
+        site_packages_path: Path,
+        next_resolver: BaseDependencyResolver | None = None,
+    ):
+        super().__init__(next_resolver)
+        self._file_loader = file_loader
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/files.py:164]: @ericvergnaud : Question from the other PR. Why does the site package resolver require a file loader?

File: src/databricks/labs/ucx/source_code/files.py [hunk 2]
```diff
@@ -150,3 +150,29 @@ def resolve_import(self, path_lookup: PathLookup, name: str) -> MaybeDependency:
...
+        self._site_packages_path = site_packages_path
+
+    def with_next_resolver(self, resolver: BaseDependencyResolver) -> BaseDependencyResolver:
+        return SitePackageResolver(self._file_loader, self._site_packages_path, resolver)
+
+    def resolve_import(self, path_lookup: PathLookup, name: str) -> MaybeDependency:
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/files.py:170]: Should this work for submodules as well? `databricks.labs.ucx`

File: src/databricks/labs/ucx/contexts/application.py [hunk 3]
```diff
@@ -358,10 +358,14 @@ def notebook_loader(self) -> NotebookLoader:
     def notebook_resolver(self):
         return NotebookResolver(self.notebook_loader)
 
+    @cached_property
+    def site_packages_path(self):
+        lookup = self.path_lookup
+        return next(path for path in lookup.paths if "site-packages" in path.as_posix())
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/contexts/application.py:364]: We returning only the first site-packages? What should we do with (possible) later ones?

File: src/databricks/labs/ucx/source_code/files.py (hunk 2 above)
Comment (JCZuurmond) [reply src/databricks/labs/ucx/source_code/files.py:170]: Should `SitePackage` not resolve a library instead? Why an import? When do we use which?

File: src/databricks/labs/ucx/source_code/files.py [hunk 4]
```diff
@@ -150,3 +150,29 @@ def resolve_import(self, path_lookup: PathLookup, name: str) -> MaybeDependency:
...
+        path = Path(self._site_packages_path, name)
+        if not path.is_dir():
+            return super().resolve_import(path_lookup, name)
+        path = Path(path, "__init__.py")
+        if not path.is_file():
+            return super().resolve_import(path_lookup, name)
+        dependency = Dependency(self._file_loader, path)
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/files.py:177]: We are not using like the PipResolver does

**Review by JCZuurmond — CHANGES_REQUESTED** [review]
@ericvergnaud : I have some clarification points

File: src/databricks/labs/ucx/source_code/files.py (hunk 4 above)
Comment (JCZuurmond) [reply src/databricks/labs/ucx/source_code/files.py:177]: I would expect a similar output

File: src/databricks/labs/ucx/contexts/application.py [hunk 5]
```diff
@@ -373,8 +377,8 @@ def file_loader(self):
         return FileLoader()
 
     @cached_property
-    def site_packages_resolver(self):
-        return SitePackageResolver(self.site_packages, self.file_loader, self.path_lookup)
+    def site_package_resolver(self):
+        return SitePackageResolver(self.file_loader, self.path_lookup, self.site_packages_path)
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/contexts/application.py:381]:
```suggestion
        return SitePackageResolver(self.file_loader, self.site_packages_path)
```

File: src/databricks/labs/ucx/source_code/files.py (hunk 2 above)
Comment (JCZuurmond) [reply src/databricks/labs/ucx/source_code/files.py:170]: Decision on naming needs to be made for import vs library

**Comment by JCZuurmond** [comment]
@ericvergnaud : Will you first merge #1686?

File: src/databricks/labs/ucx/contexts/application.py (hunk 3 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/contexts/application.py:364]: It's a good question but maybe not in scope of this PR. Requires %pip and job task settings implementation to find out.

File: src/databricks/labs/ucx/source_code/files.py (hunk 4 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/files.py:177]: Not sure I understand your comment.

File: src/databricks/labs/ucx/contexts/application.py (hunk 5 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/contexts/application.py:381]: Fixed

**Comment by ericvergnaud** [comment]
> @ericvergnaud : Will you first merge #1686? There seems to be overlap. Or you could scope #1686 to only change the framework and this PR to only fix the site package resolver logic

We need to merge this first, then #1686

File: src/databricks/labs/ucx/source_code/files.py (hunk 2 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/files.py:170]: Fixed submodule import

**Review by JCZuurmond — APPROVED** [review]
LGTM, Discussed over video chat, resolved outstanding comments

**Pull request merged**